DataFormats/MuonReco/src/MuonSelectors.cc

```diff
@@ -929,3 +949,48 @@ int muon::sharedSegments( const reco::Muon& mu, const reco::Muon& mu2, unsigned
   
     return ret; 
 }
+
+void muon::setCutBasedSelectorFlags(reco::Muon& muon, 
+				    const reco::Vertex* vertex,
+				    bool run2016_hip_mitigation)
+{
+  // https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideMuonIdRun2
+  unsigned int selectors = muon.selectors();
+  // Compute Id and Isolation variables
+  double chIso  = muon.pfIsolationR04().sumChargedHadronPt;
+  double nIso   = muon.pfIsolationR04().sumNeutralHadronEt;
+  double phoIso = muon.pfIsolationR04().sumPhotonEt;
+  double puIso  = muon.pfIsolationR04().sumPUPt;
+  double dbCorrectedIsolation = chIso + std::max( nIso + phoIso - .5*puIso, 0. ) ;
+  double dbCorectedRelIso = dbCorrectedIsolation/muon.pt();
+  double tkRelIso = muon.isolationR03().sumPt/muon.pt();
+
+  // Base selectors
+  if (muon::isLooseMuon(muon))        selectors |= reco::Muon::CutBasedIdLoose;
+  if (vertex){
+    if (muon::isTightMuon(muon,*vertex))  selectors |= reco::Muon::CutBasedIdTight;
+    if (muon::isSoftMuon(muon,*vertex,run2016_hip_mitigation))   selectors |= reco::Muon::SoftCutBasedId;
+    if (muon::isHighPtMuon(muon,*vertex)) selectors |= reco::Muon::CutBasedIdGlobalHighPt;
+    if (muon::isTrackerHighPtMuon(muon,*vertex)) selectors |= reco::Muon::CutBasedIdTrkHighPt;
```

Comment:
hi @drkovalskyi - are none of these selectors selecting a subset of another one? (i.e., tight is often a subset of muon candidates passing the loose selector)

 Then an if else would help reduce the number of muon candidates to be checked

Reply:
Most of the selectors were created at different points in times and often use quite different selections. They also often have different targets in mind optimized for different topologies. We have reduced the number of muons of fake muons in the muon collection in 92X by a factor of two by getting rid of non-arbitrated TrackerMuons, so the total number of calls per event is pretty small.

Reply:
@davidlange6 
Historically all these IDs were nod designed as strict subsets. So, if we take opportunity of incidental nesting, it may have to be undone in a less comfortable way when an improvement comes to one but not another ID.
Better keep the assignments factorized.